service/src/test/java/bio/terra/workspace/service/danglingresource/DanglingResourceCleanupServiceTest.java

```diff
@@ -0,0 +1,150 @@
+    workspace =
+        workspaceUtils.createWorkspaceWithGcpContext(userAccessUtils.defaultUserAuthRequest());
+  }
+
+  /**
+   * Delete workspace. Doing this outside of test bodies ensures cleanup runs even if tests fail.
+   */
+  @AfterEach
+  public void cleanup() {
+    workspaceService.deleteWorkspace(workspace, userAccessUtils.defaultUserAuthRequest());
+  }
+
+  @Test
+  @DisabledIfEnvironmentVariable(named = "TEST_ENV", matches = BUFFER_SERVICE_DISABLED_ENVS_REG_EX)
+  void cleanupResourcesSuppressExceptions_cleansDanglingResources_succeeds() throws Exception {
+
+    // Create a controlled AI notebook instance
+    ApiGcpAiNotebookInstanceResource notebook =
+        mockMvcUtils
+            .createAiNotebookInstance(
+                userAccessUtils.defaultUserAuthRequest(), workspace.getWorkspaceId(), null)
+            .getAiNotebookInstance();
+
+    // Create a controlled gce instance
+    ApiGcpGceInstanceResource instance =
+        mockMvcUtils
+            .createGceInstance(
+                userAccessUtils.defaultUserAuthRequest(), workspace.getWorkspaceId(), null)
+            .getGceInstance();
+
+    // Create a controlled gcs bucket
+    String bucketResourceName = TestUtils.appendRandomNumber("my-bucket");
+    String bucketCloudName = TestUtils.appendRandomNumber("my-bucket");
+    ApiGcpGcsBucketResource bucket =
+        mockMvcUtils
+            .createControlledGcsBucket(
+                userAccessUtils.defaultUserAuthRequest(),
+                workspace.getWorkspaceId(),
+                bucketResourceName,
+                bucketCloudName, /*location*/
+                null,
+                /*storageClass*/ null,
+                /*lifecycle*/ null)
+            .getGcpBucket();
+
+    // Directly delete the notebook
+    AIPlatformNotebooksCow notebooksCow = crlService.getAIPlatformNotebooksCow();
+    InstanceName instanceName =
+        InstanceName.builder()
+            .projectId(notebook.getAttributes().getProjectId())
+            .location(notebook.getAttributes().getLocation())
+            .instanceId(notebook.getAttributes().getInstanceId())
+            .build();
+    try {
+      OperationCow<Operation> deleteOperation =
+          notebooksCow
+              .operations()
+              .operationCow(notebooksCow.instances().delete(instanceName).execute());
+      GcpUtils.pollAndRetry(deleteOperation, Duration.ofSeconds(20), Duration.ofMinutes(12));
+    } catch (Exception e) {
+      throw new RuntimeException("Failed to delete notebook instance", e);
+    }
+
+    // Call dangling resource cleanup
+    danglingResourceCleanupService.cleanupResourcesSuppressExceptions();
+
+    // Verify that the notebook is deleted, polling to ensure dangling resource cleanup flight to
+    // delete the notebook has completed
+    boolean isDeleted = false;
+    for (int retryCount = 0; retryCount < ASSERT_RESOURCE_CLEANUP_RETRY_COUNT; retryCount++) {
```

Comment:
i don't like we are pulling this instead of the flight.. but it's a list of flight so either way not very clean.

Reply:
Agreed 😞